### ISSUE TITLE
Property decompilation bug

### ISSUE THREAD

**Issue by CreateAndInject** [icsharpcode/ILSpy#1159]
[Test.zip](https://github.com/icsharpcode/ILSpy/files/2065700/Test.zip)

```
private static void Main(string[] args)
{
	Second second = new Second();
	second.Info = "hello";
	Console.WriteLine(second.get_Info()); //second.Info
	Console.Read();
}
```

Reflector & dnSpy are both correct.

**Comment by CreateAndInject** [comment]
Generally speaking, ILSpy doesn't restore property when lacking some assembies, but this one doesn't lack anything.

**Comment by siegfriedpammer** [comment]
Yes, but looking at the IL:
```il
.method private hidebysig static 
	void Main (
		string[] args
	) cil managed 
{
	// Method begins at RVA 0x2074
	// Code size 38 (0x26)
	.maxstack 2
	.entrypoint
	.locals init (
		[0] class Second
	)

	IL_0000: nop
	IL_0001: newobj instance void Second::.ctor()
	IL_0006: stloc.0
	IL_0007: ldloc.0
	IL_0008: ldstr "hello"
	IL_000d: callvirt instance void First::set_Info(string)
	IL_0012: nop
	IL_0013: ldloc.0
	IL_0014: call instance string Second::get_Info()
	IL_0019: call void [mscorlib]System.Console::WriteLine(string)
	IL_001e: nop
	IL_001f: call int32 [mscorlib]System.Console::Read()
	IL_0024: pop
	IL_0025: ret
} // end of method Program::Main
```

As you can see, the instruction at 0xd calls the setter through the declaring type `First`. This is expected. The instruction at 0x14 calls the getter of the Info property through the child class and uses a `member reference` and no `method declaration` token. This is a pattern, we currently do not support.

Do you know which compiler was used to generate this assembly?

**Comment by CreateAndInject** [comment]
I find a .net 1 application without obfuscated, ILSpy doesn't restore so many properties, but there's no lacking assemblies. So I write a test, and modify its IL like that .net 1 application, it's runnable.

Maybe ILSpy should find the method from base type if it isn't exist.

**Comment by CreateAndInject** [comment]
I think ILSpy can do it like dnSpy:
If can't resolve a MethodRef to MethodDef, and the name of the MethodRef starts with "get_" or "set_", allways use name.Substring(4)

**Comment by siegfriedpammer** [comment]
I just tested with csc and vbc for .NET 1.1... both of them do not generate such code.

Of course we could try to be smart and use the usual naming convention to display unresolvable property/event accessors differently, but this could lead to invalid programs in some other cases. Another drawback: there would be no links on these properties, because there is no definition we could link to...

Anyway, this is unlikely to get fixed soon, because it requires some refactoring of the CallBuilder and a few extra methods that work without having the full definition information available.

**Comment by CreateAndInject** [comment]
Now, I know how to generate such code by VS2003.

Create a project
```
public class First
{
    public virtual string Text
    {
        get
        {
            throw new NotImplementedException();
        }
        set
        {
        }
    }
}

public class Second : First
{
}
```

Create one more project, and ref the above project
```
public class Third : Second
{
    public override string Text
    {
        get
        {
            return base.Text;
        }
        set
        {
            base.Text = value;
        }
    }
}
```

**Comment by siegfriedpammer** [comment]
Hmmm... the code you described, always calls `[other]First::get_Text()` *when compiled with C# >= 2.0*.
Which leads the code to be decompiled to `base.Text` on every line. Because there is no other property in called `Text` in `Second` the compiler sees `Text`, `this.Text` and `base.Text` as a reference to the same property getter, that is: `First.get_Text`.

I have some faint memory of Eric Lippert writing on his MSDN blog about a bug in the C# 1.x compiler, which was fixed in C# 2, where the compiler sometimes generates calls to base methods that would crash the run-time later on. (See https://blogs.msdn.microsoft.com/ericlippert/2010/03/29/putting-a-base-in-the-middle/ for more info.)

If we adapt this scenario to your example:

```
>>other.cs -> compile with csc.exe /target:library other.cs

public class First
{
    public string Text
    {
        get
        {
            return "FIRST";
        }
        set
        {
        }
    }
}

public class Second : First
{
}

----------------------
>>this.cs -> compile with csc.exe /target:exe /r:other.dll this.cs

using System;

public class Third : Second
{
	public static void Main()
	{
		new Third().Test();
	}
	
  	public void Test()
	{
		Console.WriteLine(Text);
		Console.WriteLine(this.Text);
		Console.WriteLine(base.Text);
	}
}
```

Now when you compile both sources with the *C# 1.x* compiler and execute `this.exe`, everything is fine:
> .\this.exe
FIRST
FIRST
FIRST

Now, suppose `other.dll` is a third party dll and there is an update. Your customers do not know about this and update the third party, but not your application.

change other.cs to the following and recompile *ONLY other.dll* 
```
public class First
{
    public string Text
    {
        get
        {
            return "FIRST";
        }
        set
        {
        }
    }
}

public class Second : First
{
	private new string Text { // <-- yes, this is strange, but works!!!! (new only hides the warning...)
		get {
			return "PRIVATE SECOND";
		}
		set {}
	}
}
```

Execute `this.exe` again:
> .\this.exe
>
>Unhandled Exception: System.MethodAccessException: Second.get_Text()
>   at Third.Test()
>   at Third.Main()

And there we get a runtime error. If you recompile this.cs as well, the problem disappears, because the C# 1 compiler now discovers the new method.

If we do the same thing using C# 2 or newer, there is no such problem.

I believe, we should not fix this IL level bug by decompiling it to nice-looking C#, which actually has different semantics on later versions of C#.

Also, adding a special case, if the language version is set to C# 1.0, is just too much work for too little gain, because C# 1.x has been dead for over 10 years now...

**Comment by CreateAndInject** [comment]
How can I get MemberReference from MemberReferenceExpression? It seems there's no annotation stored in MemberReferenceExpression.
I want to fix it myself, but I can't get enough information from MemberReferenceExpression

**Comment by siegfriedpammer** [comment]
We no longer use Mono.Cecil objects on the C# AST level... you can only get the TypeSystem MemberResolveResult, but that will only give you the erroneous UnkownMemberResolveResult. As I said, this is not easy to fix. This problem affects different parts of the decompiler. In particular ILReader and DecompilerTypeSystem...

**Comment by CreateAndInject** [comment]
But I can't get MemberResolveResult from MemberReferenceExpression, there's nothing stored in the annotations.

**Comment by siegfriedpammer** [comment]
You can always use the `GetResolveResult()` extension method. But in this case you will only get an ErrorResolveResult with no information at all.

**Comment by siegfriedpammer** [comment]
Yes, this is an oversight.

**Comment by CreateAndInject** [comment]
Do you fix the oversight?
If there's no problem with the lastest commit, how can I fix it on the base of Cecil version (v3.2)?
